### PR TITLE
Adversarial test: pickles step domain size too large

### PR DESCRIPTION
**Pull request by mrmr1993**
Checklist:

- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

### CODE REVIEW

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

File: src/lib/pickles/pickles.ml [hunk 1]
```diff
@@ -2998,244 +2998,322 @@ module Make_str (_ : Wire_types.Concrete) = struct
...
 
-        let dummy_proof =
-          Proof0.dummy Nat.N2.n Nat.N2.n Nat.N2.n ~domain_log2:15
+        module Recurse_on_bad_proof = struct
+          open Impls.Step
 
-        type _ Snarky_backendless.Request.t +=
-          | Proof : (Nat.N2.n, Nat.N2.n) Proof0.t Snarky_backendless.Request.t
+          let dummy_proof =
+            Proof0.dummy Nat.N2.n Nat.N2.n Nat.N2.n ~domain_log2:15
```
Comment (mimoo) [inline src/lib/pickles/pickles.ml:3240]: when I was trying to trigger the bug I believe I had to modify that as well

File: src/lib/pickles/pickles.ml [hunk 2]
```diff
@@ -2998,244 +2998,322 @@ module Make_str (_ : Wire_types.Concrete) = struct
...
-                let y =
-                  exists Field.typ ~compute:(fun () -> Field.Constant.of_int 0)
+        let rule self : _ Inductive_rule.t =
+          { identifier = "main"
+          ; prevs = [ self; self ]
+          ; main =
+              (fun { public_input = () } ->
+                let dummy_proof =
+                  As_prover.Ref.create (fun () ->
+                      Proof0.dummy Nat.N2.n Nat.N2.n Nat.N2.n ~domain_log2:15 )
```
Comment (mimoo) [inline src/lib/pickles/pickles.ml:3115]: I had to modify this one as well (domain_log2) to trigger the bug

Comment (mimoo) [reply src/lib/pickles/pickles.ml:3115]: Screenshot 2023-02-08 at 4 43 33 PM

Comment (mimoo) [reply src/lib/pickles/pickles.ml:3115]: so my functor as expecting this additional argument: 

Screenshot 2023-02-08 at 4 44 00 PM

File: src/lib/pickles/pickles.ml [hunk 3]
```diff
@@ -3313,6 +3320,20 @@ module Make_str (_ : Wire_types.Concrete) = struct
                     { stmt.proof_state.deferred_values with b = shift_value b }
                 }
             }
+
+          let check_verifier_error err =
+            (* Convert to JSON to make it easy to parse. *)
+            err |> Error_json.error_to_yojson
```
Comment (mimoo) [inline src/lib/pickles/pickles.ml:3326]: wouldn't it make more sense to use a variant to hold all types of errors? (although maybe that's too much work for this PR, but I mean in general)

**Review by mimoo — APPROVED** [review]
LGTM besides what I noted, but if you managed to trigger the error without these then I might have been wrong

**Comment by mrmr1993** [comment]
NB: This includes the changes from #12632 to avoid racing with it to merge.
Not changing the merge base because apparently that resets the reviews now :|

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

**Pull request merged**